src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -572,14 +572,57 @@ bool HandleRedirect (HttpStatusCode redirectCode, HttpURLConnection httpConnecti
 			if (redirectState.RedirectCounter >= MaxAutomaticRedirections)
 				throw new WebException ($"Maximum automatic redirections exceeded (allowed {MaxAutomaticRedirections}, redirected {redirectState.RedirectCounter} times)");
 
-			string redirectUrl = locationHeader [0];
-			string protocol = httpConnection.URL?.Protocol;
-			if (redirectUrl.StartsWith ("//", StringComparison.Ordinal)) {
-				// When redirecting to an URL without protocol, we use the protocol of previous request
-				// See https://tools.ietf.org/html/rfc3986#section-5 (example in section 5.4)
-				redirectUrl = protocol + ":" + redirectUrl;
-			}
-			redirectState.NewUrl = new Uri (redirectUrl, UriKind.Absolute);
+			Uri redirectUrl;
+			try {
+				if (Logger.LogNet)
+					Logger.Log (LogLevel.Debug, LOG_APP, $"Raw redirect location: {locationHeader [0]}");
+
+				redirectUrl = new Uri (locationHeader [0], UriKind.RelativeOrAbsolute);
+			} catch (Exception ex) {
+				throw new WebException ($"Invalid redirect URI received: {locationHeader [0]}", ex);
+			}
+
+			UriBuilder builder = null;
+			if (!redirectUrl.IsAbsoluteUri || String.IsNullOrEmpty (redirectUrl.Host)) {
+				if (Logger.LogNet) {
+					if (!redirectUrl.IsAbsoluteUri)
+						Logger.Log (LogLevel.Debug, LOG_APP, $"Relative redirect URL: {redirectUrl}");
+					else
+						Logger.Log (LogLevel.Debug, LOG_APP, $"Redirect URL without host/schema: {redirectUrl}");
+					Logger.Log (LogLevel.Debug, LOG_APP, "Building URL based on the original request URL");
+				}
+
+				// Handle relative URLs as well as absolute ones without the schema
+
+				// If the Location had an URL starting with `//`, the Uri class will report that it's an
+				// absolute URI and that it has no host (which is not true). In this case we must take
+				// over and handle the situation manually - the part following `//` and preceeding the
+				// next `/` is a host name and the scheme is taken from the original URL.
+				if (redirectUrl.AbsolutePath.StartsWith ("//", StringComparison.Ordinal)) {
+					redirectUrl = new Uri ($"{httpConnection.URL.Protocol}:{redirectUrl.AbsolutePath}", UriKind.Absolute);
+					if (Logger.LogNet)
+						Logger.Log (LogLevel.Debug, LOG_APP, $"Fixed up a scheme-less URL to: {redirectUrl}");
+				} else {
+					// When redirecting to an URL without protocol and without host, we use the protocol of previous request
+					// See https://tools.ietf.org/html/rfc3986#section-5 (example in section 5.4)
+					builder = new UriBuilder (httpConnection.URL.ToString ()) {
+						Path = redirectUrl.AbsolutePath,
+						Query = redirectUrl.Query,
+						Fragment = null,
```

Comment:
Should we be clearing `Fragment` like this?  I can't readily see which example in RFC3986 applies, if any.

Reply:
Yes, in this case we should. It may not see obvious, but  this `UriBuilder` instantiation uses `httpConnection.URL` **only** to get the base host, scheme and port and **not** to get the other parts of the **original** URI. Note that `Path` and `Query` are set from the redirect URI. The fragment is set properly in the code below  611:

```csharp
if (!String.IsNullOrEmpty (httpConnection.URL.Ref) && String.IsNullOrEmpty (redirectUrl.Fragment)) {
  if (Logger.LogNet)
    Logger.Log (LogLevel.Debug, LOG_APP, $"Appending fragment '{httpConnection.URL.Ref}' to redirect URL '{redirectUrl}'");

  if (builder == null)
    builder = new UriBuilder (redirectUrl);
  builder.Fragment = httpConnection.URL.Ref;
}
```

It may seem that line 611 should just leave `Fragment` alone considering we use `httpConnection.URL` to create the builder, but I prefer to have a clear separation of concerns here (and `builder` may have to be instantiated in other cases as well in the future - I don't want to have to change the `Fragment` setting code when/if that happens).